src/native/fs/fs_napi.cpp

```diff
@@ -213,6 +216,182 @@ struct Stat : public FSWorker
+    Getxattr(const Napi::CallbackInfo& info) 
+        : FSWorker(info)
+    {
+        Napi::Object config = info[0].As<Napi::Object>();
+        _md_prefix = config.Get("md_prefix").ToString();
+        _path = info[1].As<Napi::String>();
+        Begin(XSTR() << "Getxattr " << DVAL(_path));
+    }
+    virtual void Work()
+    {
+        // Serves as a "lock" mechanism when reading the xattrs of a file
+        // Since this is non atomic operation, we abort if ctime of file changed
+        struct stat stat_res;
+        auto r = stat(_path.c_str(), &stat_res);
+        if (r) {
+            SetSyscallError();
+            return;
+        }
+        auto start_ctime = stat_res.st_ctime;
+
+        #ifdef __APPLE__
+            ssize_t buf_len = listxattr(_path.c_str(), NULL, 0, 0);
+        #else
+            ssize_t buf_len = listxattr(_path.c_str(), NULL, 0);
+        #endif
+
+        // No xattr, nothing to do  
+        if (buf_len == 0) return;
+
+        if (buf_len == -1) {
+            SetSyscallError();
+            return;
+        }
+
+        Buf buf(buf_len);
+        if (buf.cdata() == NULL) {
+            SetError(XSTR() << "Getxattr: buf allocation failure");
+            return;
+        }
+
+        #ifdef __APPLE__
+            buf_len = listxattr(_path.c_str(), buf.cdata(), buf_len, 0);
+        #else
+            buf_len = listxattr(_path.c_str(), buf.cdata(), buf_len);
+        #endif
+
+        if (buf_len == -1) {
+            SetSyscallError();
+            return;
+        }
+
+        while (buf_len > 0) {
+            std::string key(buf.cdata());
+
+            // Get only prefixed keys
+            if (key.rfind(_md_prefix, 0) != 0) {
+                buf_len -= key.size() + 1;
+                buf.slice(key.size() + 1, buf_len);
+                continue;
+            }
+
+            #ifdef __APPLE__
+                ssize_t value_len = getxattr(_path.c_str(), key.c_str(), NULL, 0, 0, 0);
+            #else
+                ssize_t value_len = getxattr(_path.c_str(), key.c_str(), NULL, 0);
+            #endif
+
+            if (value_len == -1) {
+                SetSyscallError();
+                return;
+            }
+
+            if (value_len > 0) {
+                Buf val(value_len + 1);
+                if (val.cdata() == NULL) {
+                    SetError(XSTR() << "Getxattr: val allocation failure");
+                    return;
+                }
+
+                #ifdef __APPLE__
+                    value_len = getxattr(_path.c_str(), key.c_str(), val.cdata(), value_len, 0, 0);
+                #else
+                    value_len = getxattr(_path.c_str(), key.c_str(), val.cdata(), value_len);
+                #endif
+
+                if (value_len == -1) {
+                    SetSyscallError();
+                    return;
+                }
+                
+                val[value_len] = 0;
+                _md[key] = std::string(val.cdata());
```

Comment:
send the length to string ctor and then you don't need to add null terminator and it won't need to find its length on its own:
```suggestion
                _md[key] = std::string(val.cdata(), value_len);
```